### PR TITLE
sink: add barrier in underlying sink to support accruate stop

### LINKED ISSUE #2230: CDC cloud: data inconsistency after CDC owner switching
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

- Pause 1x changefeed
- Load data to upstream tidb: $ bin/go-ycsb load mysql -P workloads/betting -p recordcount=9900000 -p mysql.host=xxx -p mysql.port=4000 --threads 200 -p dbnameprefix=testcc  -p databaseproportions=1.0 -p unitnameprefix=unit1 -p unitscount=1 -p tablecount=200 -p loadbatchsize=500 -p mysql.password=12345678
- Make CDC owner switch: $ kubectl delete pod db-ticdc-1 --kubeconfig=/etc/kubernetes/cluster1.conf -n tidbxxx,  delete ticdc-0, ticdc-1, ticdc-2 one by one.
- Load data to upstream again:  $ bin/go-ycsb run mysql -P workloads/betting -p operationcount=5000000 -p mysql.host=xxx -p mysql.port=4000 --threads 200 -p dbnameprefix=testcc  -p databaseproportions=1.0 -p unitnameprefix=unit1 -p unitscount=1 -p tablecount=200  -p mysql.password=12345678
- Make all pd restart:  $ kubectl delete pod db-pd-0 db-pd-1 db-pd-2  --kubeconfig=/etc/kubernetes/cluster1.conf -n tidbxxx

- Check changefeed status, when upstream has stopped to write, and sync completed, check if data consistent.

2. What did you expect to see?

- All data consistecy.

3. What did you see instead?

- 11 tables in 200 tables have data inconsistency, rows count are the same, but some value are different.
sync-diff log:
[2021/07/06 09:54:56.487 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_12`] [where="((TRUE) AND TRUE)"] ["source checksum"=3733393904] ["target checksum"=2605381674] ["get source checksum cost"=250.960066ms] ["get target checksum cost"=206.317042ms]
[2021/07/06 09:55:05.426 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_18`] [where="((TRUE) AND TRUE)"] ["source checksum"=3258024002] ["target checksum"=317619229] ["get source checksum cost"=183.408826ms] ["get target checksum cost"=87.105012ms]
[2021/07/06 09:55:21.657 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_11`] [where="((TRUE) AND TRUE)"] ["source checksum"=446946590] ["target checksum"=3514961352] ["get source checksum cost"=214.137411ms] ["get target checksum cost"=146.654629ms]
[2021/07/06 09:55:45.994 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_14`] [where="((TRUE) AND TRUE)"] ["source checksum"=4122421317] ["target checksum"=274903492] ["get source checksum cost"=209.464336ms] ["get target checksum cost"=109.468714ms]
[2021/07/06 09:56:19.416 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_79`] [where="((TRUE) AND TRUE)"] ["source checksum"=1531020405] ["target checksum"=3060465391] ["get source checksum cost"=78.444038ms] ["get target checksum cost"=8.828986ms]
[2021/07/06 09:56:31.858 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_1`] [where="((TRUE) AND TRUE)"] ["source checksum"=1478455333] ["target checksum"=1441817492] ["get source checksum cost"=538.675238ms] ["get target checksum cost"=434.188737ms]
[2021/07/06 09:57:15.840 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_statistics_agent_game_day`] [where="((TRUE) AND TRUE)"] ["source checksum"=3721642887] ["target checksum"=4238525603] ["get source checksum cost"=82.975369ms] ["get target checksum cost"=19.861323ms]
[2021/07/06 09:57:19.254 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_21`] [where="((TRUE) AND TRUE)"] ["source checksum"=4167676186] ["target checksum"=3207274853] ["get source checksum cost"=116.469711ms] ["get target checksum cost"=107.899528ms]
[2021/07/06 09:57:25.812 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_61`] [where="((TRUE) AND TRUE)"] ["source checksum"=1741577261] ["target checksum"=3192701895] ["get source checksum cost"=76.139605ms] ["get target checksum cost"=15.00056ms]
[2021/07/06 09:57:35.276 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_statistics_user_game_day`] [where="((TRUE) AND TRUE)"] ["source checksum"=3141552518] ["target checksum"=4014241092] ["get source checksum cost"=90.358791ms] ["get target checksum cost"=30.339078ms]
[2021/07/06 09:57:50.224 +00:00] [WARN] [diff.go:551] ["checksum is not equal"] [table=`testcc0`.`unit10_game_bets_game_tag_analysis_31`] [where="((TRUE) AND TRUE)"] ["source checksum"=1131104107] ["target checksum"=2463968134] ["get source checksum cost"=139.692822ms] ["get target checksum cost"=86.867459ms]

For exmaple in table testcc0. unit10_statistics_agent_game_day, rows count is 4155, the different row is " id: 946 “

4. Versions of the cluster

    - Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

        ```
        5.0.2
        ```

    - TiCDC version (execute `cdc version`):

        ```
        ["Welcome to Change Data Capture (CDC)"] [release-version=v5.0.0-dev] [git-hash=9300b05ceec2c4811198416a5d21e9f4910deddd] [git-branch=cloud-cdc-5.0] [utc-build-time="2021-06-26 14:16:57"] [go-version="go version go1.16.3 linux/amd64"] [failpoint-build=false]
        ```

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve? 

Fix #2230 

This PR introduces a new `Barrier` API in sink interface, which is a synchronous function to wait all events to be flushed in underlying sink, note once `Barrier` is called, the resolved ts can't be pushed until the Barrier call returns.

We call `Barrier` when table sink is closing (via the `Close` method of table sink), and before `Barrier` is called, we draw back all existing events in sink buffer.

Note the cherry-pick cannot be merged into release-4.0, release-5.0 until new owner is enabled in these branches.

### What is changed and how it works?

- The `Barrier` implementation
   - MySQL sink: wait checkpoint-ts to reach max received resolved-ts
   - MQ sink: TODO
- Drawback mechanism in buffer sink
   - Buffer sink stores rows by table in memory, and emit rows to underlying sink only resolved ts is forwarded.
   - Hook a channel from sink manager to notify buffer sink to draw back rows of a table, and wait callback

### Check List 

Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
Fix a bug that multiple processors could write the same table when this table is re-scheduling
```

### CODE REVIEW

File: cdc/owner/ddl_puller.go [hunk 1]
```diff
@@ -119,7 +120,12 @@ func (h *ddlPullerImpl) Run(ctx cdcContext.Context) error {
 		}
 		h.mu.Lock()
 		defer h.mu.Unlock()
+		if job.ID == h.lastDDLJobID {
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply cdc/owner/ddl_puller.go]: has been removed

File: cdc/changefeed.go [hunk 2]
```diff
@@ -1009,8 +1009,10 @@ func (c *changeFeed) Close() {
 		}
 	}
 
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply cdc/changefeed.go:1013]: added

File: cdc/owner.go [hunk 3]
```diff
@@ -437,7 +437,7 @@ func (o *Owner) newChangeFeed(
 	}
 	defer func() {
 		if resultErr != nil && primarySink != nil {
-			primarySink.Close()
+			primarySink.Close(ctx)
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply cdc/owner.go:440]: added comment

**Comment by amyangfei** [comment]
ptal @liuzix @overvenus

File: cdc/sink/mysql.go [hunk 4]
```diff
@@ -861,13 +869,53 @@ func (w *mysqlSinkWorker) cleanup() {
...
+			if s.checkpointTs() >= maxResolvedTs {
+				return nil
+			}
+			checkpointTs, err := s.FlushRowChangedEvents(ctx, maxResolvedTs)
+			if err != nil {
+				return err
+			}
+			if checkpointTs >= maxResolvedTs {
+				return nil
+			}
```
Comment (overvenus) [inline cdc/sink/mysql.go:903]: Do we need to add sleep here to avoid hogging CPU?

File: cdc/sink/manager.go [hunk 5]
```diff
@@ -167,28 +178,43 @@ func (t *tableSink) EmitCheckpointTs(ctx context.Context, ts uint64) error {
...
+	drawbackChan chan drawbackMsg,
+) Sink {
 	sink := &bufferSink{
-		Sink: backendSink,
-		buffer: make(chan struct {
-			rows       []*model.RowChangedEvent
-			resolvedTs model.Ts
-		}, defaultBufferChanSize),
+		Sink:         backendSink,
+		buffer:       make(map[model.TableID][]*model.RowChangedEvent),
```
Comment (overvenus) [inline cdc/sink/manager.go:214]: Does it have a capacity uppper limit? Will it cause OOM?

Comment (amyangfei) [reply cdc/sink/manager.go:214]: It shares the same flow control as table sink, added a comment

File: cdc/sink/mysql.go (hunk 4 above)
Comment (amyangfei) [reply cdc/sink/mysql.go:903]: addressed

File: cdc/sink/mysql.go [hunk 6]
```diff
@@ -861,13 +869,55 @@ func (w *mysqlSinkWorker) cleanup() {
...
+func (s *mysqlSink) Barrier(ctx context.Context) error {
+	warnDuration := 3 * time.Minute
+	ticker := time.NewTicker(warnDuration)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case <-ticker.C:
+			ticker.Reset(warnDuration)
```
Comment (ben1009) [inline cdc/sink/mysql.go:888]: stupid question, why reset, thx

Comment (amyangfei) [reply cdc/sink/mysql.go]: not needed, removed

**Review by overvenus — APPROVED** [review]
LGTM

**Review by ben1009 — APPROVED** [review]

**Pull request merged**